src/dbvt/mod.rs

```diff
@@ -0,0 +1,1667 @@
+/// It is recommended to use the latter when possible. If the former is used,
+/// [`reindex_values`](struct.DynamicBoundingVolumeTree.html#method.reindex_values)
+/// must be called if the order of the values is changed in any way.
+///
+/// # Type parameters:
+///
+/// - `T`: A type that implements [`TreeValue`](trait.TreeValue.html), and is usable in the tree.
+///        Needs to be able to store the node index of itself, and handle its own bound and
+///        fattened bound.
+///
+/// # Examples
+///
+/// ```
+/// # extern crate cgmath;
+/// # extern crate collision;
+///
+/// use cgmath::{Point2, Vector2, InnerSpace};
+/// use collision::{Aabb, Aabb2, Ray2};
+/// use collision::dbvt::{DynamicBoundingVolumeTree, TreeValue, ContinuousVisitor};
+///
+/// #[derive(Debug, Clone)]
+/// struct Value {
+///     index: usize,
+///     pub id: u32,
+///     pub aabb: Aabb2<f32>,
+///     fat_aabb: Aabb2<f32>,
+/// }
+///
+/// impl Value {
+///     pub fn new(id: u32, aabb: Aabb2<f32>) -> Self {
+///         Self {
+///             index: 0,
+///             id,
+///             fat_aabb : aabb.add_margin(Vector2::new(3., 3.)),
+///             aabb,
+///         }
+///     }
+/// }
+///
+/// impl TreeValue for Value {
+///     type Vector = Vector2<f32>;
+///     type Bound = Aabb2<f32>;
+///
+///     fn bound(&self) -> &Aabb2<f32> {
+///         &self.aabb
+///     }
+///
+///     fn fat_bound(&self) -> Aabb2<f32> {
+///         self.fat_aabb.clone()
+///     }
+///
+///     fn set_index(&mut self, index: usize) {
+///         self.index = index
+///     }
+///
+///     fn index(&self) -> usize {
+///         self.index
+///    }
+/// }
+///
+/// fn aabb2(minx: f32, miny: f32, maxx: f32, maxy: f32) -> Aabb2<f32> {
+///    Aabb2::new(Point2::new(minx, miny), Point2::new(maxx, maxy))
+/// }
+///
+/// fn main() {
+///     let mut tree = DynamicBoundingVolumeTree::<Value>::new();
+///     tree.insert(Value::new(10, aabb2(5., 5., 10., 10.)));
+///     tree.insert(Value::new(11, aabb2(21., 14., 23., 16.)));
+///     tree.do_refit();
+///
+///     let ray = Ray2::new(Point2::new(0., 0.), Vector2::new(-1., -1.).normalize());
+///     let visitor = ContinuousVisitor::<Ray2<f32>, Value>::new(&ray);
+///     assert_eq!(0, tree.query(&visitor).len());
+///
+///     let ray = Ray2::new(Point2::new(6., 0.), Vector2::new(0., 1.).normalize());
+///     let visitor = ContinuousVisitor::<Ray2<f32>, Value>::new(&ray);
+///     let results = tree.query(&visitor);
+///     assert_eq!(1, results.len());
+///     assert_eq!(10, results[0].0.id);
+///     assert_eq!(Point2::new(6., 5.), results[0].1);
+/// }
+/// ```
+///
+pub struct DynamicBoundingVolumeTree<T: TreeValue> {
+    nodes: Vec<Node<T::Bound>>,
+    values: Vec<T>,
+    free_list: Vec<usize>,
+    updated_list: Vec<usize>,
+    root_index: usize,
+    refit_nodes: Vec<(u32, usize)>,
+}
+
+impl<T: TreeValue> Debug for DynamicBoundingVolumeTree<T>
```

Comment:
it's not recommended to mix the bounds within `impl<>` and `where` blocks. Let's just move everything into `where` here

Reply:
Ah, did that slip through. I decided to do a custom Debug for the tree, that spits out a dot parseable graph. I shall fix.